File: examples/python/quickstart.py
#!/usr/bin/env python3
"""
FaaS Platform Python Quick Start Examples

Demonstrates basic usage patterns and common operations.
"""

import asyncio
import sys
import os
# Fix the SDK path - use the correct location
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../sdks/python')))

from faas_sdk import FaaSClient, Runtime, ExecutionMode


async def main():
    # Initialize client
    client = FaaSClient("http://localhost:8080")

    print("🚀 FaaS Platform Python Examples\n")

    # Example 1: Simple Python execution
    print("1. Running Python code:")
    result = await client.run_python('print("Hello from Python!")')
    print(f"   Output: {result.output}")
    print(f"   Duration: {result.duration_ms}ms")
    print(f"   Cache hit: {result.cache_hit}\n")

    # Example 2: JavaScript execution
    print("2. Running JavaScript code:")
    result = await client.run_javascript('console.log("Hello from Node.js!")')
    print(f"   Output: {result.output}")
    print(f"   Duration: {result.duration_ms}ms\n")

    # Example 3: Bash script execution
    print("3. Running Bash script:")
    result = await client.run_bash('''
        echo "System info:"
        uname -a
        echo "Memory:"
        free -h | head -2
    ''')
    print(f"   Output:\n{result.output}\n")

    # Example 4: Using Docker runtime explicitly
    print("4. Using Docker runtime:")
    result = await client.execute(
        command='echo "Running in Docker container"',
        runtime=Runtime.DOCKER
    )
    print(f"   Output: {result.output}")
    print(f"   Runtime used: {result.runtime_used}\n")

    # Example 5: Using environment variables
    print("5. With environment variables:")
    result = await client.execute(
        command='python -c "import os; print(f\'API_KEY={os.environ.get(\'API_KEY\', \'not set\')}\')"',
        image="python:3.11-slim",
        env_vars={"API_KEY": "secret123"}
    )
    print(f"   Output: {result.output}\n")

    # Example 6: Caching demonstration
    print("6. Caching demonstration:")

    # First execution (cold)
    result1 = await client.run_python('import time; time.sleep(1); print("Computed result")')
    print(f"   First run: {result1.duration_ms}ms (cache hit: {result1.cache_hit})")

    # Second execution (should be cached)
    result2 = await client.run_python('import time; time.sleep(1); print("Computed result")')
    print(f"   Second run: {result2.duration_ms}ms (cache hit: {result2.cache_hit})")

    if result2.duration_ms < result1.duration_ms / 10:
        print("   ✅ Caching working! Second run was much faster\n")

    # Example 7: Pre-warming containers
    print("7. Pre-warming containers:")
    await client.prewarm("python:3.11-slim", count=3)
    print("   Pre-warmed 3 Python containers for instant execution\n")

    # Example 8: Error handling
    print("8. Error handling:")
    try:
        result = await client.run_python('import sys; sys.exit(1)')
        if result.error:
            print(f"   Error caught: {result.error}\n")
    except Exception as e:
        print(f"   Exception: {e}\n")

    # Example 9: Getting metrics
    print("9. Platform metrics:")

    # Server metrics
    server_metrics = await client.get_metrics()
    print(f"   Server metrics: {server_metrics}")

    # Client metrics
    client_metrics = client.get_client_metrics()
    print(f"   Client metrics:")
    print(f"     Total requests: {client_metrics.total_requests}")
    print(f"     Cache hit rate: {client_metrics.cache_hit_rate:.2%}")
    print(f"     Avg latency: {client_metrics.average_latency_ms:.2f}ms\n")

    # Example 10: Health check
    print("10. Platform health:")
    health = await client.health_check()
    print(f"    Status: {health['status']}")
    print(f"    Components: {health['components']}")

    await client.session.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/quickstart.ts
#!/usr/bin/env ts-node
/**
 * FaaS Platform TypeScript Quick Start Examples
 *
 * Demonstrates basic usage patterns and common operations.
 */

import { FaaSClient, Runtime, ExecutionMode } from '../../sdks/typescript/src';

async function main() {
  // Initialize client
  const client = new FaaSClient('http://localhost:8080');

  console.log('🚀 FaaS Platform TypeScript Examples\n');

  // Example 1: Simple Python execution
  console.log('1. Running Python code:');
  let result = await client.runPython('print("Hello from Python!")');
  console.log(`   Output: ${result.output}`);
  console.log(`   Duration: ${result.durationMs}ms`);
  console.log(`   Cache hit: ${result.cacheHit}\n`);

  // Example 2: JavaScript execution
  console.log('2. Running JavaScript code:');
  result = await client.runJavaScript('console.log("Hello from Node.js!")');
  console.log(`   Output: ${result.output}`);
  console.log(`   Duration: ${result.durationMs}ms\n`);

  // Example 3: TypeScript execution
  console.log('3. Running TypeScript code:');
  result = await client.runTypeScript(`
    interface Person {
      name: string;
      age: number;
    }
    const person: Person = { name: "Alice", age: 30 };
    console.log(\`Hello, \${person.name}!\`);
  `);
  console.log(`   Output: ${result.output}\n`);

  // Example 4: Bash script execution
  console.log('4. Running Bash script:');
  result = await client.runBash(`
    echo "System info:"
    uname -a
    echo "Node version:"
    node --version 2>/dev/null || echo "Node not available"
  `);
  console.log(`   Output:\n${result.output}\n`);

  // Example 5: Using Docker runtime explicitly
  console.log('5. Using Docker runtime:');
  result = await client.execute({
    command: 'echo "Running in Docker container"',
    runtime: Runtime.Docker
  });
  console.log(`   Output: ${result.output}`);
  console.log(`   Runtime used: ${result.runtimeUsed}\n`);

  // Example 6: Using environment variables
  console.log('6. With environment variables:');
  result = await client.execute({
    command: 'node -e "console.log(`API_KEY=${process.env.API_KEY}`)"',
    image: 'node:20-slim',
    envVars: { API_KEY: 'secret123' }
  });
  console.log(`   Output: ${result.output}\n`);

  // Example 7: Caching demonstration
  console.log('7. Caching demonstration:');

  // First execution (cold)
  const start1 = Date.now();
  const result1 = await client.runJavaScript(`
    const start = Date.now();
    while(Date.now() - start < 1000) {} // Simulate work
    console.log("Computed result");
  `);
  console.log(`   First run: ${result1.durationMs}ms (cache hit: ${result1.cacheHit})`);

  // Second execution (should be cached)
  const result2 = await client.runJavaScript(`
    const start = Date.now();
    while(Date.now() - start < 1000) {} // Simulate work
    console.log("Computed result");
  `);
  console.log(`   Second run: ${result2.durationMs}ms (cache hit: ${result2.cacheHit})`);

  if (result2.durationMs < result1.durationMs / 10) {
    console.log('   ✅ Caching working! Second run was much faster\n');
  }

  // Example 8: Pre-warming containers
  console.log('8. Pre-warming containers:');
  await client.prewarm('node:20-slim', 3);
  console.log('   Pre-warmed 3 Node.js containers for instant execution\n');

  // Example 9: Error handling
  console.log('9. Error handling:');
  try {
    result = await client.runJavaScript('throw new Error("Intentional error")');
    if (result.error) {
      console.log(`   Error caught: ${result.error}\n`);
    }
  } catch (error) {
    console.log(`   Exception: ${error}\n`);
  }

  // Example 10: Event listeners
  console.log('10. Event listeners:');
  client.on('execution', (event) => {
    console.log(`   Execution event: ${event.elapsedMs}ms, cached: ${event.cacheHit}`);
  });

  client.on('retry', (event) => {
    console.log(`   Retry attempt ${event.attempt}: ${event.error}`);
  });

  await client.runPython('print("Testing events")');

  // Example 11: Getting metrics
  console.log('\n11. Platform metrics:');

  // Server metrics
  const serverMetrics = await client.getMetrics();
  console.log(`   Server metrics:`, serverMetrics);

  // Client metrics
  const clientMetrics = client.getClientMetrics();
  console.log(`   Client metrics:`);
  console.log(`     Total requests: ${clientMetrics.totalRequests}`);
  console.log(`     Cache hit rate: ${(clientMetrics.cacheHitRate * 100).toFixed(2)}%`);
  console.log(`     Avg latency: ${clientMetrics.averageLatencyMs.toFixed(2)}ms\n`);

  // Example 12: Health check
  console.log('12. Platform health:');
  const health = await client.healthCheck();
  console.log(`    Status: ${health.status}`);
  console.log(`    Components:`, health.components);

  // Example 13: Method chaining
  console.log('\n13. Method chaining:');
  const dockerClient = new FaaSClient('http://localhost:8080')
    .useDocker()
    .setCaching(true);

  result = await dockerClient.runPython('print("Using Docker with caching")');
  console.log(`   Output: ${result.output}`);

  const firecrackerClient = new FaaSClient('http://localhost:8080')
    .useFirecracker()
    .setCaching(false);

  try {
    result = await firecrackerClient.runPython('print("Using Firecracker VMs")');
    console.log(`   Output: ${result.output}`);
  } catch (error) {
    console.log('   Firecracker not available on this system');
  }
}

// Run the examples
main()
  .then(() => {
    console.log('\n✅ All examples completed successfully!');
    process.exit(0);
  })
  .catch((error) => {
    console.error('\n❌ Error running examples:', error);
    process.exit(1);
  });